### DISCUSSION TITLE
Migration from walletconnect_dart to walletconnect_flutter_v2

### DISCUSSION THREAD

**Discussion by nqhhdev** [WalletConnect/walletconnect-monorepo#2603] [Developer Support] [answered]
I need to support to migrationnn plss

**Comment by ratio91** [comment, marked as answer]
you are not alone - many users of the `walletconnect_dart` package are currently facing the same task.
A proper migration guide could help a lot of flutter devs. Anything planned here?

**Reply by nqhhdev** [reply]
IMO: I need to update walletconnect_v2, and docs WC_v2 is very hard for me. I need support

**Comment by finessevanes** [comment]
@nqhhdev @ratio91 Hey all, thanks for the feedback on the docs. We have documentation available for v2 Flutter. Could you please let us know which places you find to be lacking context?

**Comment by ratio91** [comment]
@finessevanes specifically the changes from v1 to v2 are hard to keep track of.

A separate guide on the most important functions and how they changed from v1 (walletconnect_dart) to v2 would be super helpful!
*) how to add the connection of a dapp to a wallet (e.g. metamask)  
*) how to access the connected account(s) easily  
*) best practises how to wrap it in a riverpod provider, etc.

**Reply by nqhhdev** [reply]
I agreee with your pointttt

**Comment by nqhhdev** [comment]
@finessevanes  Thanks for your support ❤️

- First thing: Let's do it in the easiest way, for example a person new to DApp can use it for the most basic purpose: Connect DApp, session management, signing...
- The new docs are in my opinion quite hard to understand. I need it clean and clear

- I have a question for you: `Walletconnect_dart` still supported after Jun 28?

**Reply by finessevanes** [reply]
>I have a question for you: Walletconnect_dart still supported after Jun 28?

Nope!

As for the other questions/suggestions, I've pinged our flutter team for clarirty.

**Comment by ratio91** [comment]
I found this guide quite helpful: [https://medium.com/walletconnect/walletconnect-sign-v2-0-beginners-guide-for-javascript-developers-c02c02d215c9](https://medium.com/walletconnect/walletconnect-sign-v2-0-beginners-guide-for-javascript-developers-c02c02d215c9)
It is for JS, but it can be used in a similar way in flutter.
A blog post like this for the flutter package would be super helpful for many people. Maybe someone of the walletconnect team can find the time ;-)

**Comment by nqhhdev** [comment]
@finessevanes 
- Can you support us or not?
- Help us answer questions

**Comment by nqhhdev** [comment]
@finessevanes Any update about it??

**Reply by finessevanes** [reply]
Sharing w/ the team! Thanks for your fb

**Comment by bobwith2bees** [comment]
It's not a migration guide, but feel free to play with my example app:  https://gitlab.com/graflr/flutter_web3_demo
I will get back to working on it as soon as I get my own WalletConnect V1 EOL mitigation done for my app.  Be gentle on any feedback. ;-)

Migration notes off the top of my head:

* The API is more configurable, which makes doing some of the basics - frankly a bit harder. 
* The walletconnect registry can help query wallets that are compatible with your platform.  There are updated query parameters that assist with V2 capabilities.   I think the URL might have changed as well. 
* Chain ID's... read up on CAIP (https://github.com/ChainAgnostic/CAIPs/blob/master/README.md).    It is no longer an EVM compatible world.  I had to rewrite my fundamental blockchain and account classes.  Great when you want to go multichain. If you just want a single chain...  it's a bit of a bummer.  (See example app model/blockchain and blockchain account files.)
* The WC URI now has brackets in the query parameter.   This can cause encoding issues that you might not have seen in V1. 
```
flutter: 12:25:10 💡 ExplorerRegSvc  - launchApp - 1. mobile universal using LaunchMode.externalApplication 
https://keyring.app//wc?uri=wc:a1f88723ca5c8feb55ed356d5317894d3d847d7f73fdf30aa34109f571e8e929@2?relay-protocol=irn&symKey=e984f3cd611ad9364a74200e60750bde764ef20439bf01a4f899b74dadbb2ec5&methods=[wc_sessionPropose,wc_sessionRequest]
```
* If your app has permission to multiple accounts/chains, you no longer have to ask the wallet to switch accounts/chains.  (I THINK this is true...)  But the wallet can change it's mind and give you a new list of accounts/wallets at any time.  
* The pairing session will timeout in 30 days.  The Flutter client does not appear to extend the session. https://github.com/WalletConnect/WalletConnectFlutterV2/issues/55 
* Don't even try to use the session or paring pings. https://github.com/WalletConnect/WalletConnectFlutterV2/issues/85
* Keyring Pro works with V2.  Other wallets I am getting results all over the place. It could me my code (or that bracket encoding...?)
I am still not confident on the mobile launching (universal / native) working as expected.
* Learn the difference between a pairing topic and session topic
* Similar to V1 - you are communicating via a Relay.  Monitor the relay health, so you don't mistake a relay issue for an app issue.
* Clear out everything on your app as well as the wallet when testing.    If you app or wallet crashes, there may not be automatic clean up of relay connections.  So if you are trying to create a new connection - there is no programmatic way to see if the connection is a duplicate (and perhaps deny it.)   The last thing you need in your UI is three connections to the same wallet/DAPP for the same device. 
https://github.com/WalletConnect/WalletConnectFlutterV2/issues/83



Example debug
```
flutter: 12:25:08 🐛 WalletConnctSvc - createWalletConnectSession - Create SignClient connection request for eip155:1,eip155:137. isRelayClientUp: true, isRelayClientUp: false
flutter: 12:25:08 🐛 WalletConnctSvc - createWalletConnectSession - requiredNamespace: { chains: [eip155:1, eip155:137], methods:[eth_signTypedData, personal_sign, eth_signTransaction], events: [accountsChanged, chainChanged, connect, disconnect] }
flutter: 12:25:09 🐛 WalletConnctSvc - createWalletConnectSession - paringUri: wc:a1f88723ca5c8feb55ed356d5317894d3d847d7f73fdf30aa34109f571e8e929@2?relay-protocol=irn&symKey=e984f3cd611ad9364a74200e60750bde764ef20439bf01a4f899b74dadbb2ec5&methods=[wc_sessionPropose,wc_sessionRequest]
flutter: 12:25:10 🐛 WalletConnctSvc - createWalletConnectSession - Returned from showWalletSelectionDialog with selectedRegistryListing: KEYRING PRO
 mobile    native: keyring://
 mobile universal: https://keyring.app/
         app_type: wallet
flutter: 12:25:10 💡 ExplorerRegSvc  - launchApp - 1. mobile universal using LaunchMode.externalApplication 
https://keyring.app//wc?uri=wc:a1f88723ca5c8feb55ed356d5317894d3d847d7f73fdf30aa34109f571e8e929@2?relay-protocol=irn&symKey=e984f3cd611ad9364a74200e60750bde764ef20439bf01a4f899b74dadbb2ec5&methods=[wc_sessionPropose,wc_sessionRequest]
flutter: 12:25:10 🐛 HomePage        - AppLifecycleState: AppLifecycleState.inactive.
flutter: 12:25:10 🐛 WalletConnctSvc - createWalletConnectSession - launchAppWithParingUri returned true
flutter: 12:25:10 🐛 WalletConnctSvc - Awaiting session completer before returning.
flutter: 12:25:18 🐛 HomePage        - AppLifecycleState: AppLifecycleState.paused.
flutter: 12:25:31 🐛 WalletConnctSvc - createWalletConnectSession - Connection with wallet took 22752.07 seconds.
flutter: 12:25:31 🐛 WalletConnctSvc - =Core Services wc: v2, relayUrl: wss://relay.walletconnect.org, pushUrl: https://echo.walletconnect.com, uptime: 58 seconds)=
flutter: 12:25:31 🐛 WalletConnctSvc - ==Pairings (2)==
flutter: 12:25:31 🐛 WalletConnctSvc -      2:28 topic: 732efe7769ee3be117d88e990957bebf3ed8305f9a44c0135ca8ebdc803ae2fa, relay irn, active: false peer: <unavailable>
flutter: 12:25:31 🐛 WalletConnctSvc - 30.0000 days topic: a1f88723ca5c8feb55ed356d5317894d3d847d7f73fdf30aa34109f571e8e929, relay irn, active: true peer: Keyring
flutter: 12:25:31 🐛 WalletConnctSvc - =Sign Client Info - (wc v2)=
flutter: 12:25:31 🐛 WalletConnctSvc - ==Sign Client Sessions (4)==
flutter: 12:25:31 🐛 WalletConnctSvc - <expired> topic: 1e42bd532b09d1d6343e99393abeadeb5cd65b17228913fe5c6eb44cee068618, pairing topic: b14f7b8e50eb5ba47cb0894e1c8b103e9bd5d3a07997873b963a6f9e9b88ec0c ack: true, peer name: 🌈 Rainbow accounts: [ eip155:1:0xf12e82E33245f3cDbA5dA82Ea0599431144594Cf, eip155:1:0x7Fd611Af09BeFB5f47aBB19FCC25a22997aEb0b1 ]
flutter: 12:25:31 🐛 WalletConnctSvc - <expired> topic: 34330c5b0487e750f5dad0986818d805d8aca9928f47ad941f5eb62ec3de6686, pairing topic: ce35c0ecece0a59a7a102628e2e5057f306c0e5354e4e6d136b6b75129a92249 ack: true, peer name: Spot accounts: [ eip155:1:0x7Fd611Af09BeFB5f47aBB19FCC25a22997aEb0b1 ]
flutter: 12:25:31 🐛 WalletConnctSvc - <expired> topic: 4fb860d3a73146abe8281041f8913b6fb046c756cdc10759928f28eeecbf12d0, pairing topic: 04b45231e1885c4e171325166bd8abbcec25c993f1dc78aaf320e893314e8c29 ack: true, peer name: BitKeep accounts: [ eip155:1:0xE45d31e7E28AFC211a6046EA5D25AE93E687d1e3, eip155:137:0xE45d31e7E28AFC211a6046EA5D25AE93E687d1e3 ]
flutter: 12:25:31 🐛 WalletConnctSvc - 7.0000 days topic: 2518935b49dba43bdb7199177d12ee00b3ee0504d9c33aeb4754658cb9b1b30f, pairing topic: a1f88723ca5c8feb55ed356d5317894d3d847d7f73fdf30aa34109f571e8e929 ack: true, peer name: Keyring accounts: [ eip155:1:0x0a639a9dd60d699c2cb4219082e26a4180b5263d, eip155:137:0x0a639a9dd60d699c2cb4219082e26a4180b5263d ]
flutter: 12:25:31 🐛 WalletConnctSvc - ==Sign Client Proposals (14)==
flutter: 12:25:31 🐛 WalletConnctSvc - <expired> topic: 9289f5251b8746b71d8eb95dbf399953f823ca752e413050182e86829b9e5144 - required: [ eip155:1, eip155:137 ], optional: [  ], session properties: null, id: 1683736470015614, relays: [ irn ]
flutter: 12:25:31 🐛 WalletConnctSvc - <expired> topic: e63aa4d522ffc1cc917142ab4c2ec7128a059fc1b7bf012ab4325f51661d7442 - required: [ eip155:1 ], optional: [  ], session properties: null, id: 1684269990350862, relays: [ irn ]
flutter: 12:25:31 🐛 WalletConnctSvc - <expired> topic: ec6d98ecc84057c9bee167157a5c6c4ee68b09980a08f3de3b5b0b367bf7c0c1 - required: [ eip155:1 ], optional: [  ], session properties: null, id: 1684270031097907, relays: [ irn ]
flutter: 12:25:31 🐛 WalletConnctSvc - <expired> topic: ffac10982cf6ec1e8ff24b0eee98a3ac9149e9b5c55009c3d0cf4d24adae5fc2 - required: [ eip155:1 ], optional: [  ], session properties: null, id: 1684270073307474, relays: [ irn ]
flutter: 12:25:31 🐛 WalletConnctSvc - <expired> topic: ddc7911a440c1ec3c800237adde05497abe79a8174fe6afd6ba0a5d0d6afda45 - required: [ eip155:1 ], optional: [  ], session properties: null, id: 1684270104921236, relays: [ irn ]
flutter: 12:25:31 🐛 WalletConnctSvc - <expired> topic: 824882ca018496f3ec475a394f2ec764962ac975a23b8e27faa9bb8d912ddd3f - required: [ eip155:1 ], optional: [  ], session properties: null, id: 1684270160211811, relays: [ irn ]
flutter: 12:25:31 🐛 WalletConnctSvc - <expired> topic: b7665ed1ebda8f78c68b55f9a3099570a9e67e85dc7de2b8cd0a01dd7005b8e1 - required: [ eip155:1 ], optional: [  ], session properties: null, id: 1684271207511998, relays: [ irn ]
flutter: 12:25:31 🐛 WalletConnctSvc - ==Sign Client Stats==
flutter: 12:25:31 🐛 WalletConnctSvc - No data.
flutter: 12:25:33 🐛 WalletConnctSvc - signEngine.onSessionConnect - 4 active sessions.
flutter: 12:25:33 💡 WalletConnctSvc - New session from Keyring - accounts: [ eip155:1:0x0a639a9dd60d699c2cb4219082e26a4180b5263d, eip155:137:0x0a639a9dd60d699c2cb4219082e26a4180b5263d ]
flutter: 12:25:33 🐛 WalletConnctSvc - 7.0000 days topic: 2518935b49dba43bdb7199177d12ee00b3ee0504d9c33aeb4754658cb9b1b30f, pairing topic: a1f88723ca5c8feb55ed356d5317894d3d847d7f73fdf30aa34109f571e8e929 ack: true, peer name: Keyring accounts: [ eip155:1:0x0a639a9dd60d699c2cb4219082e26a4180b5263d, eip155:137:0x0a639a9dd60d699c2cb4219082e26a4180b5263d ]
flutter: 12:25:34 🐛 HomePage        - AppLifecycleState: AppLifecycleState.inactive.
flutter: 12:25:34 🐛 HomePage        - AppLifecycleState: AppLifecycleState.resumed.
flutter: 12:27:36 🐛 WalletConnctSvc - Core.onRelayClientConnect - Marking relay connection up.

```

**Comment by bobwith2bees** [comment]
Few more notes for migration:
* While the registry/cloud explorer allows you to dynamically query for the latest compatible wallets - your fastest approach is still to have an explicit allow/deny list and validate wallets for your dapp one at a time.   If you go the dynamic route you need to check for:
** Missing Peer metadata.  Some can be null some, like the icons can also be an empty list.  So your UI will need fallbacks
** Registry Links are not always formatted correctly.   It is getting better but not all native links have :// at the end.  So you need to add code to reformat as needed.    Also I highly consider having a security check of the URI - these links should be considered untrusted.
** Registry listing is not guaranteed to be in sync with the app capabilities.  So a wallet might add support for a new chain, but forget to update their registry. 
*  And yes - you have to encode the walletconnect URI explicitly before you embed it into a native or universal link.  KeyringPro seemed to look for both encoded and non encoded cases (robust implementation) but others - not so much.  
* An launching just the wc intent, seems to be the wrong way to go on mobile.  Use native, fallback to universal, fallback to appstore etc.   With all the incompatibilities, it has been hard to get a clean launch function that works across the platforms.
